backend/oasst_backend/prompt_repository.py

```diff
@@ -488,3 +496,178 @@ def close_task(self, post_id: str, allow_personal_tasks: bool = False):
+
+    def fetch_post_descendants(self, message: Post | UUID, max_depth: int = None) -> list[Post]:
+        if isinstance(message, UUID):
+            message = self.fetch_post(message)
+
+        desc = self.db.query(Post).filter(Post.thread_id == message.thread_id, Post.depth > message.depth)
+        if max_depth is not None:
+            desc = desc.filter(Post.depth <= max_depth)
+
+        desc = desc.all()
+
+        return self.trace_descendants(message, desc)
+
+    def fetch_longest_conversation(self, message: Post | UUID) -> list[Post]:
+        tree = self.fetch_message_tree(message)
+        max_message = max(tree, key=lambda m: m.depth)
+        return self.trace_conversation(tree, max_message)
+
+    def fetch_message_with_max_children(self, message: Post | UUID) -> tuple[Post, list[Post]]:
+        tree = self.fetch_message_tree(message)
+        max_message = max(tree, key=lambda m: m.children_count)
+        return max_message, [m for m in tree if m.parent_id == max_message.id]
+
+    def query_messages(
+        self,
+        user_id: Optional[UUID] = None,
+        username: Optional[str] = None,
+        api_client_id: Optional[UUID] = None,
+        desc: bool = True,
+        limit: Optional[int] = 10,
+        start_date: Optional[datetime.datetime] = None,
+        end_date: Optional[datetime.datetime] = None,
+        only_roots: bool = False,
+        deleted: Optional[bool] = None,
+    ) -> list[Post]:
+        if not self.api_client.trusted and not api_client_id:
+            # Let unprivileged api clients query their own messages without api_client_id being set
+            api_client_id = self.api_client.id
+
+        if not self.api_client.trusted and api_client_id != self.api_client.id:
+            # Unprivileged api client asks for foreign messages
+            raise OasstError("Forbidden", OasstErrorCode.API_CLIENT_NOT_AUTHORIZED, HTTP_403_FORBIDDEN)
+
+        messages = self.db.query(Post)
+        if user_id:
+            messages = messages.filter(Post.person_id == user_id)
+        if username:
+            messages = messages.join(Person)
+            messages = messages.filter(Person.username == username)
+        if api_client_id:
+            messages = messages.filter(Post.api_client_id == api_client_id)
+
+        if start_date:
+            messages = messages.filter(Post.created_date >= start_date)
+        if end_date:
+            messages = messages.filter(Post.created_date < end_date)
+
+        if only_roots:
+            messages = messages.filter(Post.parent_id.is_(None))
+
+        if deleted is not None:
+            messages = messages.filter(Post.deleted == deleted)
+
+        if desc:
+            messages = messages.order_by(Post.created_date.desc())
+        else:
+            messages = messages.order_by(Post.created_date.asc())
+
+        if limit is not None:
+            messages = messages.limit(limit)
+
+        # TODO: Pagination could be great at some point
```

Comment:
the problem with pagination is always that it is only really useful with for messages .. otherwise which of the 10000 pages does the user really want to see? Also it puts stress on the database to find correct elements of visible page based on counts which means queries everything before the elements that are on the page etc. Therefore I think in many cases it is actually easier to step by name or date incrementally forward backward - this allows to page forward + backward but not to specify a page index .. but you could jump by prefix or start/end-date...

Reply:
This comment is just a reminder, not a commitment. It is definitely a subject to discussion, no pagination is intended at this moment. 
Stepping using date is just another way to paginate. I believe there are ways to introduce pagination without putting much stress on the db. On the other hand loading arbitrary many entries from the table could affect performance as well.